local types = require("../types")

export type Array<T> = types.Array<T>

return {
    all = require("./all"),
    at = require("./at"),
	concat = require("./concat"),
	every = require("./every"),
    fill = require("./fill"),
	filter = require("./filter"),
	find = require("./find"),
	findIndex = require("./findIndex"),
	flat = require("./flat"),
	flatMap = require("./flatMap"),
	forEach = require("./forEach"),
	from = require("./from"),
    fromFn = require("./fromFn"),
	includes = require("./includes"),
	indexOf = require("./indexOf"),
	isArray = require("./isArray"),
    isEmpty = require("./isEmpty"),
	join = require("./join"),
    lastIndexOf = require("./lastIndexOf"),
	map = require("./map"),
    pop = require("./pop"),
	reduce = require("./reduce"),
	shift = require("./shift"),
	slice = require("./slice"),
	some = require("./some"),
	sort = require("./sort"),
	sort_other = require("./sort_other"),
	splice = require("./splice"),
    toString = require("./toString"),
	unshift = require("./unshift"),
	push = require("./push"),
	reverse = require("./reverse"),
	spread = require("./spread"),
}